src/cpp/server/server_cc.cc

```diff
@@ -1201,6 +1119,30 @@ int Server::AddListeningPort(const grpc::string& addr,
   return port;
 }
 
+void Server::Ref() {
+  shutdown_refs_outstanding_.fetch_add(1, std::memory_order_relaxed);
+}
+
+void Server::UnrefWithPossibleNotify() {
+  if (shutdown_refs_outstanding_.fetch_add(-1, std::memory_order_acq_rel) ==
+      1) {
+    // No refs outstanding means that shutdown has been initiated and no more
+    // callback requests are outstanding.
+    grpc::internal::MutexLock lock(&mu_);
+    GPR_ASSERT(shutdown_);
+    shutdown_done_ = true;
+    shutdown_done_cv_.Signal();
+  }
+}
+
+void Server::UnrefAndWaitLocked() {
+  if (shutdown_refs_outstanding_.fetch_add(-1, std::memory_order_acq_rel) ==
+      1) {
+    shutdown_done_ = true;
```

Comment:
when we set `shutdown_done_` to true, can we just return?  Is there is a reason to call `WaitUntil()`?

Reply:
Great point, done!

Reply:
Thank you.